tls/s2n_record_write.c

```diff
@@ -94,6 +90,9 @@ S2N_RESULT s2n_record_max_write_payload_size(struct s2n_connection *conn, uint16
         }
     }
 
+    /* Ensure that we don't reserve too little space.  */
+    RESULT_ENSURE(*max_fragment_size >= S2N_TLS_MINIMUM_FRAGMENT_RESERVE_LENGTH, S2N_ERR_FRAGMENT_LENGTH_TOO_SMALL);
+
```

Comment:
Are we sure this function is actually the right place for this check? s2n_record_max_write_payload_size isn't just used to ensure sufficient buffer space. It's also used to fragment non-alert data like application data and handshake messages. Maybe you should be making this change in s2n_should_flush, where this specifically applies to ensuring sufficient space for the next record?

Reply:
I'm reasonably certain it is the right place for the check. Let's take a the places its used to fragment non-alert data.

###  Application Data
In `s2n_record_writev` we use `s2n_record_max_write_payload_size` in the calculation of the length of a fragment. We take the minimum of the size of the payload, `to_write`, and the `max_write_payload_size` (computed from `s2n_record_max_write_payload_size`). `s2n_record_write` is sometimes passed `TLS_ALERT`s to potentially fragment and send. Therefore if `s2n_record_max_write_payload_size` returns a value less than 2 we will end up fragmenting those `TLS_ALERT` despite having a `to_write==2`.

This does mean that Application data won't be fragmented in sizes smaller than 2 (unless there is only 1 byte to send). But that makes sense -- we want to make use of all the space we allocated in the out buffer.

In `s2n_sendv_with_offset_impl` we calculate the size of a record to generate for `TLS_APPLICATION_DATA`. (That is the `to_write` value in `s2n_record_writev`). It has a code path in which it sets `to_write` explictly to 1. In that case `s2n_record_writev` will write a length 1 record (because MIN(1, 2) == 1).

The question becomes do we want to force records to fragment application data to length 1, even though we had to allocate space for a record with fragment length 2? If we do want to do that then we would need to special case TLS_ALERT in s2n_record_writev to prevent it alone from being fragmented. I submit that would be overly complicated and not what we want anyway.

### TLS Handshake
In `s2n_handshake_write_io` we use `MIN(s2n_stuffer_data_available(&conn->handshake.io), max_payload_size)` (where `s2n_record_max_write_payload_size` computes the `max_payload_size`) to calculate the size to pass to `s2n_record_write`. That ends up being the `to_write` variable in `s2n_record_writev`. 

In this case the question becomes would we like to under utilize the space we had to allocate? -- Should be be requiring a lower minimum fragment size (1) than we allow the user to set (2) for handshake messages?

### The send buffer
It is clear that s2n_should_flush needs to reserve at least enough space for an alert message or the remaining data.

### The meaning of `s2n_record_max_write_payload_size`
My understanding of the `s2n_record_max_write_payload_size` is that it represents the maximum size of a payload s.t. you can put that payload in a record and store it in our internally allocated buffers. Indeed in `s2n_record_writev` it is used as the minimum when allocating the connection out buffer. 

### Could it be clearer?
I think we should keep this code the way it is -- but perhaps there is some alternative I haven't thought of that would make what is going on clearer?

Reply:
To be clear, I don't think this is super important. Nobody should be using 1 OR 2 byte fragments :)

[s2n_should_flush](https://github.com/aws/s2n-tls/blob/af80f6b47a17bab71cfdc5377e3d1af7530ba540/tls/s2n_send.c#L50): Required here. We're using the method to figure out whether the next send would succeed in all cases, and the next case could be an alert and require two bytes.

[s2n_send](https://github.com/aws/s2n-tls/blob/af80f6b47a17bab71cfdc5377e3d1af7530ba540/tls/s2n_send.c#L147): We're fragmenting application data. If the max payload size calculation did produce "1", that wouldn't be an error. Application data could be fragmented to that size.

[s2n_record_write](https://github.com/aws/s2n-tls/blob/56cd2c8f29b8a49f5a87bff85f7fde1bd838a545/tls/s2n_record_write.c#L281): The max payload calculation producing "1" would only be an error if we were writing an alert message. In all other cases, it would be fine.

[s2n_handshake_io](https://github.com/aws/s2n-tls/blob/7c3950f106b6c2e95d0ab18d194a4a6076cbaa00/tls/s2n_handshake_io.c#L988): We're fragmenting handshake messages. If the max payload size calculation did produce "1", that wouldn't be an error. Handshake messages could be fragmented to that size.

So in most cases, this check would be unnecessarily restrictive. It would only really be necessary in s2n_should_flush, where we could do the check outside of s2n_record_max_write_payload_size. The restriction that the buffer must be able to contain at least a fragment size of 2 would otherwise cover the standard alert s2n_record_write case without the need for another check.